### PR TITLE
fix werkzeug vulnerability issue

### PR DESCRIPTION
**Pull request by leotizzei**
the goal of this PR is to fix the vulnerability described in issue https://github.com/Open-EO/openeo-python-driver/issues/243. 

Main changes:
- upgrade werkzeug version to greater than or equal to 2.3.8 (`setup.py` module)
- fix unit tests that were failing due to werkzeug upgrade (`tests/test_views.py` module)

### CODE REVIEW

**Comment by soxofaan** [comment]
thank you for this PR, some notes:

- there are still failing tests in our internal CI setup 2 in `tests/test_views` and 9 in `tests/util/test_logging`
- you completely reformatted test_views.py, please do not do that in a PR unrelated to code formatting, that makes it practically impossible to review the real changes

**Comment by soxofaan** [comment]
great, thanks for the quick formatting update!

**Comment by soxofaan** [comment]
can you reproduce the failing tests, or should I share the failures in some way?

**Comment by leotizzei** [comment]
thanks for the feedback, @soxofaan ! I cannot reproduce the errors in`tests/test_views` and `tests/util/test_logging`. All these tests are passing in my local environment.

**Comment by leotizzei** [comment]
@soxofaan , I don't have access to the logs so I don't know why the checks have failed. Could you please share with me the failures?

**Comment by soxofaan** [comment]
Sorry for the delay. It's a bit hard to extract the info in sharable way, so here some samples.

failing tests:
```
    tests.test_views.TestLogging.test_user_id_logging[1.0.0]
    tests.test_views.TestLogging.test_user_id_logging[1.1.0]
    tests.util.test_logging.test_filter_flask_request_correlation_id_logging
    tests.util.test_logging.test_filter_flask_request_correlation_id_logging_overriding
    tests.util.test_logging.test_filter_flask_user_id_logging
    tests.util.test_logging.test_setup_logging_capture_warnings
    tests.util.test_logging.test_setup_logging_stderr_json
    tests.util.test_logging.test_setup_logging_file_json
    tests.util.test_logging.test_setup_logging_rotating_file_json
    tests.util.test_logging.test_setup_logging_capture_threading_exceptions
    tests.util.test_logging.test_setup_logging_capture_unhandled_exceptions
```

I think it all comes down to the same problem, 
and this test seems to have most condensed failure message:
`tests.util.test_logging.test_filter_flask_user_id_logging`:
```
def test_filter_flask_user_id_logging():
        with enhanced_logging(format="[%(user_id)s] %(message)s", context=LOGGING_CONTEXT_FLASK) as logs:
            app = flask.Flask(__name__)
            log = logging.getLogger(__name__)
    
>           log.info("Setting up app")

tests/util/test_logging.py:84: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib64/python3.8/logging/__init__.py:1434: in info
    self._log(INFO, msg, args, **kwargs)
/usr/lib64/python3.8/logging/__init__.py:1577: in _log
    self.handle(record)
/usr/lib64/python3.8/logging/__init__.py:1587: in handle
    self.callHandlers(record)
/usr/lib64/python3.8/logging/__init__.py:1649: in callHandlers
    hdlr.handle(record)
/usr/lib64/python3.8/logging/__init__.py:946: in handle
    rv = self.filter(record)
/usr/lib64/python3.8/logging/__init__.py:807: in filter
    result = f.filter(record)
openeo_driver/util/logging.py:321: in filter
    setattr(record, self.LOG_RECORD_ATTR, self.get_user_id())
openeo_driver/util/logging.py:316: in get_user_id
    if flask._app_ctx_stack.top:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

name = '_app_ctx_stack'

    def __getattr__(name: str) -> t.Any:
        if name == "__version__":
            import importlib.metadata
            import warnings
    
            warnings.warn(
                "The '__version__' attribute is deprecated and will be removed in"
                " Flask 3.1. Use feature detection or"
                " 'importlib.metadata.version(\"flask\")' instead.",
                DeprecationWarning,
                stacklevel=2,
            )
            return importlib.metadata.version("flask")
    
>       raise AttributeError(name)
E       AttributeError: _app_ctx_stack

venv38/lib64/python3.8/site-packages/flask/__init__.py:60: AttributeError
```

File: openeo_driver/util/logging.py [hunk 1]
```diff
@@ -313,8 +313,9 @@ def set_user_id(cls, user_id: str):
     @classmethod
     def get_user_id(cls) -> Union[str, None]:
         """Get user id as stored in Flask request global `g`."""
-        if flask._app_ctx_stack.top:
+        if flask.g:
```
Comment (soxofaan) [inline openeo_driver/util/logging.py:316]: This change probably requires a minimum flask version? Would be good to set this lower bound in setup.py

Comment (leotizzei) [reply openeo_driver/util/logging.py:316]: I'm not sure if it is necessary to set a minimum flask version, given that we had already set a minimum werkzeug version, which is one of the flask's dependencies. The oldest flask version that supports werkzeug>=2.3.8 is flask 3.0.0 (which requires werkzeug>=3.0.0 - please see https://github.com/pallets/flask/blob/735a4701d6d5e848241e7d7535db898efb62d400/pyproject.toml). But I don't have a strong opinion on that and it is fine to me if we change the requirements to restrict both werkzeug and flask. 

I think that we should set the minimum werkzeug version to 3.0.1 and not 2.3.8, as werkzeug version 3.0.0 is affected by the vulnerability (https://github.com/pallets/werkzeug/security/advisories/GHSA-hrfv-mqp8-q5rw)

Comment (leotizzei) [reply openeo_driver/util/logging.py:316]: @soxofaan ,  please let me know if there is anything else I should implement in order to merge this PR?

Comment (soxofaan) [reply openeo_driver/util/logging.py:316]: Apologies, I'm just back from holidays, catching up. I hope I can give this a final review later this week

Note that some depending projects (like openeo-aggregator) currently assume flask 2 so I have to cross-check that bumping these kind of constraints doesn't break things downstream

**Comment by soxofaan** [comment]
FYI: I'm going to cherrypick from this PR and do the constraint bumping in steps: first to `werkzeug>=2.3.8,<3.0.0` to allow upstream projects to get their setup in shape too.

**Comment by soxofaan** [comment]
werkzeug is now bumped on master to `werkzeug>=2.3.8,<3.0.0`  (549a82a4eb5183f5c01679c7da23c26be5544cf8)

**Pull request closed without merging**